File: internal/digger/src/digger/crypto/des.cpp
#include <digger/crypto/des.h>

using namespace Digger;
using namespace Digger::Crypto;

template <size_t Size>
Bits ByteToBits(Byte byte) {
    Bits bits(Size);

    for (size_t i = 0; i < Size; i += 1) {
        bits[i] = (byte >> (Size - i - 1)) & 1;
    }

    return bits;
}

template <size_t Size>
Byte BitsToByte(const Bits& bits) {
    Byte byte = 0;

    for (size_t i = 0; i < Size; i += 1) {
        byte = byte | (bits[i] << (Size - i - 1));
    }

    return byte;
}

Bits BytesToBits(const Bytes& bytes) {
    Bits bits(8 * bytes.size());

    auto it = bits.begin();

    for (auto byte : bytes) {
        auto part = ByteToBits<8>(byte);

        it = std::copy(part.begin(), part.end(), it);
    }

    return bits;
}

Bytes BitsToBytes(const Bits& bits, size_t size) {
    Bytes bytes(size);

    auto it = bits.begin();
    auto part = Bits(8);

    for (size_t i = 0; i < size; i += 1) {
        std::copy(it, it + 8, part.begin());
        it += 8;

        bytes[i] = BitsToByte<8>(part);
    }

    return bytes;
}

std::pair<Bits, Bits> SplitBits(const Bits& bits, size_t size) {
    Bits first(size);
    Bits second(size);

    std::copy(bits.begin(), bits.begin() + size, first.begin());
    std::copy(bits.begin() + size, bits.end(), second.begin());

    return std::pair(first, second);
}

Bits JoinBits(const Bits& first, const Bits& second) {
    Bits bits(first.size() + second.size());

    auto it = bits.begin();

    it = std::copy(first.begin(), first.end(), it);
    it = std::copy(second.begin(), second.end(), it);

    return bits;
}

Bits Xor(const Bits& bits1, const Bits& bits2) {
    auto size = std::min(bits1.size(), bits2.size());

    Bits bits(size);

    for (size_t i = 0; i < size; i += 1) {
        bits[i] = bits1[i] ^ bits2[i];
    }

    return bits;
}

template <size_t Size>
Bits Permutate(const Bits& bits, const Table<Size>& table) {
    Bits result(Size);

    for (size_t i = 0; i < Size; i += 1) {
        result[i] = bits[table[i] - 1];
    }

    return result;
}

template <size_t Size>
Bits Expand(const Bits& bits, const Table<Size>& table) {
    Bits result(Size);

    for (size_t i = 0; i < Size; i += 1) {
        result[i] = bits[table[i] - 1];
    }

    return result;
}

template <size_t Size>
Bits Substitute(const Bits& bits, const Mapping<Size>& mapping) {
    Bytes pieces(8);

    auto it = bits.begin();
    Bits part(6);

    for (size_t i = 0; i < 8; i += 1) {
        std::copy(it, it + 6, part.begin());
        it += 6;

        pieces[i] = BitsToByte<6>(part);
    }

    Bits result(4 * 8);
    auto it2 = result.begin();

    for (size_t i = 0; i < 8; i += 1) {
        auto piece = pieces[i];

        auto row = (piece & 1) | ((piece >> 4) & 0b10);
        auto column = (piece & 0b011110) >> 1;

        auto value = static_cast<Byte>(mapping[i][row * 16 + column]);
        auto part = ByteToBits<4>(value);

        it2 = std::copy(part.begin(), part.end(), it2);
    }

    return result;
}

Bits DES::Function(const Bits& bits, const Bits& key) const {
    Bits result;

    result = Expand(bits, Ctx.Expansion);
    result = Xor(result, key);
    result = Substitute(result, Ctx.Sbox);
    result = Permutate(result, Ctx.RoundPermutation);

    return result;
}

KeySchedule DES::GenerateKeySchedule(const Bytes& key, size_t rounds) const {
    Bits bits = BytesToBits(key);

    bits = Permutate(bits, Ctx.PC1);
    auto pair = SplitBits(bits, 28);

    KeySchedule schedule;

    for (size_t round = 0; round < rounds; round += 1) {
        auto rotation = Ctx.KeyRotation[round];

        std::rotate(pair.first.begin(), pair.first.begin() + rotation, pair.first.end());
        std::rotate(pair.second.begin(), pair.second.begin() + rotation, pair.second.end());

        auto key = JoinBits(pair.first, pair.second);
        key = Permutate(key, Ctx.PC2);

        schedule.push_back(key);
    }

    return schedule;
}

Bytes DES::ProcessBlock(const Bytes& text, const KeySchedule& schedule) const {
    auto bits = BytesToBits(text);

    bits = Permutate(bits, Ctx.InitialPermutation);
    auto pair = SplitBits(bits, 32);

    for (auto key : schedule) {
        auto result = Function(pair.second, key);
        auto tmp = Xor(result, pair.first);

        pair.first = pair.second;
        pair.second = tmp;
    }

    bits = JoinBits(pair.second, pair.first);
    bits = Permutate(bits, Ctx.FinalPermutation);

    return BitsToBytes(bits, 8);
}

DES::DES(const Context& ctx, const Bytes& key, size_t rounds)
    : Ctx(ctx) {
    RoundKeys = GenerateKeySchedule(key, rounds);
    ReversedRoundKeys = KeySchedule(RoundKeys.rbegin(), RoundKeys.rend());
}

DES::~DES() { }

Bytes DES::Encrypt(const Bytes& plaintext) const {
    return ProcessBlock(plaintext, RoundKeys);
}

Bytes DES::Decrypt(const Bytes& ciphertext) const {
    return ProcessBlock(ciphertext, ReversedRoundKeys);
}

File: checkers/digger/des.py
#!/usr/bin/env python3

from typing import List, Tuple

import tables


Bits = List[int]
Bytes = bytes

Table = Tuple[int]
Mapping = List[Tuple[int]]

KeySchedule = List[Bits]

 
def int_to_bits(value: int, size: int) -> Bits:
    bits = []

    for i in reversed(range(size)):
        bits.append((value >> i) & 1)

    return bits


def bits_to_int(bits: Bits) -> int:
    value = 0

    for bit in bits:
        value <<= 1
        value |= bit

    return value


def block_to_bits(block: Bytes) -> Bits:
    bits = []

    for byte in block:
        bits.extend(int_to_bits(byte, 8))

    return bits


def bits_to_block(bits: Bits, size: int) -> Bytes:
    block_bytes = []

    for i in range(size):
        byte = bits_to_int(bits[i * 8 : (i + 1) * 8])
        block_bytes.append(byte)

    return Bytes(block_bytes)


def xor(bits1: Bits, bits2: Bits) -> Bits:
    return [x ^ y for x, y in zip(bits1, bits2)]


def permutate(bits: Bits, table: Table) -> Bits:
    return [
        bits[table[i] - 1]
        for i in range(len(table))
    ]


def expand(bits: Bits, table: Table) -> Bits:
    return [
        bits[table[i] - 1]
        for i in range(len(table))
    ]


def substitute(bits: Bits, mapping: Mapping) -> Bits:
    pieces = []

    for i in range(8):
        piece_bits = bits[i * 6 : (i + 1) * 6]

        piece = bits_to_int(piece_bits)
        pieces.append(piece)

    values = []

    for i, piece in enumerate(pieces):
        row = (piece & 1) | ((piece >> 4) & 0b10)
        column = (piece & 0b011110) >> 1

        values.append(mapping[i][row * 16 + column])

    result_bits = []

    for value in values:
        result_bits.extend(int_to_bits(value, 4))

    return result_bits


class DES:
    def __init__(self, key: Bytes, rounds: int = 16):
        self.rounds = rounds
        self.round_keys = self._key_schedule(key)
        self.reversed_round_keys = self.round_keys[::-1]

    def encrypt(self, block: Bytes) -> Bytes:
        return self._process_block(block, self.round_keys)

    def decrypt(self, block: Bytes) -> Bytes:
        return self._process_block(block, self.reversed_round_keys)

    def _key_schedule(self, key: Bytes) -> KeySchedule:
        key = block_to_bits(key)

        key = permutate(key, tables.PC1)
        left, right = key[:28], key[28:]

        round_keys = []

        for i in range(self.rounds):
            rotation = tables.KEY_ROTATION[i]

            left = left[rotation:] + left[:rotation]
            right = right[rotation:] + right[:rotation]

            round_key = permutate(left + right, tables.PC2)
            round_keys.append(round_key)

        return round_keys

    def _process_block(self, block: Bytes, schedule: KeySchedule) -> Bytes:
        bits = block_to_bits(block)

        bits = permutate(bits, tables.INITIAL_PERMUTATION)
        left, right = bits[:32], bits[32:]

        for i in range(self.rounds):
            new_right = self._function(right, schedule[i])
            left, right = right, xor(new_right, left)

        result_bits = right + left
        result_bits = permutate(result_bits, tables.FINAL_PERMUTATION)

        return bits_to_block(result_bits, 8)

    def _function(self, bits: Bits, key_bits: Bits) -> Bits:
        bits = expand(bits, tables.EXPANSION)
        bits = xor(bits, key_bits)
        bits = substitute(bits, tables.SBOX)
        bits = permutate(bits, tables.ROUND_PERMUTATION)

        return bits


def main():
    des = DES(b'12345678')
    assert des.encrypt(b'0A1B2C3D') == b'X\xa1K\xcek\x0c\xc4\x90'
    assert des.decrypt(b'X\xa1K\xcek\x0c\xc4\x90') == b'0A1B2C3D'


if __name__ == '__main__':
    main()
